src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java

```diff
@@ -48,9 +51,59 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
             } else {
                 cart.getItems().add(new CartItem(product, quantity));
             }
+            recalculateCart(cart);
         }
     }
 
+    @Override
+    public void updateProductInCart(Cart cart, Long productId, int quantity) throws OutOfStockException, ProductNotFoundException {
+        Object syncObject = SyncObjectPool.getSyncObject(cart.getId().toString());
+        synchronized (syncObject) {
+            Product product = productService.getProductById(productId);
+            Optional<CartItem> optionalCartItem = findCartItem(cart, product);
+            checkQuantity(product, quantity, 0);
+
+            if (optionalCartItem.isPresent()) {
+                if (optionalCartItem.get().getQuantity() == quantity) {
+                    throw new SameArgumentException(optionalCartItem.get());
+                    //it is for resource economy: if the quantity is the same there is no sense to waste our resources
```

Comment:
it is in update method, in "add" we still can add the same product. When we click button "update" all the products are going to be updated. With the same quantity too: like we had 1 product and now it is still 1, and it will be updated and i suppose it is not so optimized. For example if it will be database there is no need to access it. So i decided to create this custom exception to show if the products are really updated or not, and if none of products are updated there won't be message "successfully updated"